src/dlsnode/storage/BucketFile.d

```diff
@@ -334,6 +343,70 @@ public class BucketFile: OutputStream
         }
     }
 
+
+    /**************************************************************************
+
+        Opens the bucket file and reads or writes the bucket header in a
+        non-blocking fashion.
+
+        Params:
+            path = file path
+            suspended_job = JobNotification to block
+                the fiber on until read is completed
+            file_buffer = buffer used for buffered input
+            style = style in which file will be open for
+
+    **************************************************************************/
+
+    public bool openAsync (cstring path,
+            JobNotification suspended_job,
+            void[] file_buffer,
+            File.Style style = File.ReadExisting)
+    in
+    {
+        assert(style == File.ReadExisting || style == File.ReadWriteAppending,
+                "Currently BucketFile only supports ReadExisting or ReadWriteAppending");
+    }
+    body
+    {
+        if (!this.is_open_ || style == File.ReadWriteAppending)
+        {
+            this.promise.reset();
+            this.buffered_input.reset(file_buffer);
+            this.file.open(path, style);
+            this.file_length_ = this.file.length;
+            this.file_pos_ = this.file.position;
+            this.is_open_ = true;
+        }
+
+        // In case we opened this file for writting,
+        // but the position is at 0 (empty file), we need to
+        // write the magic header.
+        if (style == File.ReadWriteAppending && this.file_length_ == 0)
+        {
+            this.writeBucketHeader();
```

Comment:
Is this blocking or non-blocking?

Reply:
This is interesting one - writes are nonblocking from the PoV of the application - the only thing application is doing is that it copies the data into the memory and then there's something in kernel that's called writeback daemon which transfers the dirty pages to the disk.

Reply:
Perhaps the comment is due here?

Reply:
Ok, that's what I was guessing. I suppose for someone who's familiar with this framework, that's obvious? Could be worth a mini comment, though, for the aid of those learning about it.

Reply:
This is a interesting trivia, really - the behaviour of the Linux system is that all the writes to the file chunks are already in page cache are really nonblocking. It doesn't have to do much with the framework. I'll add a comment here.